sql/logictest: TestParallel failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/2cbfb514fed209e9e4192bd07af6baa8dd073bab

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestParallel PKG=github.com/cockroachdb/cockroach/pkg/sql/logictest TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=971481&tab=buildLog

```
I181017 05:59:34.503063 5003 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n3: context canceled
I181017 05:59:34.507936 4942 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n4: context canceled
I181017 05:59:34.508547 4943 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n5: context canceled
W181017 05:59:34.511164 4908 internal/client/txn.go:532  [n3,replicate,s3,r31/2:/System/tsd/cr.node.s{ql…-ys…}] failure aborting transaction: node unavailable; try another peer; abort caused by: log-range-event: failed to send RPC: sending to all 5 replicas failed; last error: (err: node unavailable; try another peer) <nil>
E181017 05:59:34.511633 4908 storage/queue.go:791  [n3,replicate,s3,r31/2:/System/tsd/cr.node.s{ql…-ys…}] change replicas of r31 failed: log-range-event: failed to send RPC: sending to all 5 replicas failed; last error: (err: node unavailable; try another peer) <nil>
I181017 05:59:34.512399 4988 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n2: context canceled
W181017 05:59:34.513979 2385 storage/store.go:1504  [n5,s5,r1/3:/{Min-System/}] could not gossip first range descriptor: node unavailable; try another peer
W181017 05:59:34.519251 1853 storage/raft_transport.go:583  [n3] while processing outgoing Raft queue to node 1: rpc error: code = Canceled desc = grpc: the client connection is closing:
I181017 05:59:34.524744 4995 util/stop/stopper.go:537  quiescing; tasks left:
4      [async] closedts-subscription
1      node.Node: writing summary
1      [async] closedts-rangefeed-subscriber
I181017 05:59:34.526088 4995 util/stop/stopper.go:537  quiescing; tasks left:
3      [async] closedts-subscription
1      node.Node: writing summary
1      [async] closedts-rangefeed-subscriber
W181017 05:59:34.526393 3507 storage/raft_transport.go:583  [n3] while processing outgoing Raft queue to node 4: EOF:
I181017 05:59:34.530508 4995 util/stop/stopper.go:537  quiescing; tasks left:
1      node.Node: writing summary
I181017 05:59:34.548830 4996 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] closedts-rangefeed-subscriber
W181017 05:59:34.554558 1412 storage/store.go:1504  [n2,s2,r1/5:/{Min-System/}] could not gossip first range descriptor: node unavailable; try another peer
I181017 05:59:34.556283 4969 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n2: context canceled
I181017 05:59:34.556599 4968 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n3: context canceled
W181017 05:59:34.593558 1663 server/node.go:887  [n3,summaries] health alerts detected: {Alerts:[{StoreID:3 Category:METRICS Description:queue.replicate.process.failure Value:1}]}
W181017 05:59:34.597665 1985 ts/db.go:198  [n4,ts-poll] node unavailable; try another peer
W181017 05:59:34.601144 1412 storage/store.go:1504  [n2,s2,r1/5:/{Min-System/}] could not gossip first range descriptor: node unavailable; try another peer
I181017 05:59:34.605556 4991 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n3: context canceled
W181017 05:59:34.605797 1663 server/node.go:858  [n3,summaries] error recording status summaries: node unavailable; try another peer
W181017 05:59:34.620572 2385 storage/store.go:1504  [n5,s5,r1/3:/{Min-System/}] could not gossip first range descriptor: node unavailable; try another peer
I181017 05:59:34.652888 5030 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n2: context canceled
I181017 05:59:34.653379 5029 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n1: context canceled
I181017 05:59:34.654884 5031 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n4: context canceled
I181017 05:59:34.656531 5032 rpc/nodedialer/nodedialer.go:91  [ct-client] unable to connect to n5: context canceled
I181017 05:59:34.768654 25 sql/logictest/parallel_test.go:267  1 out of 2 parallel tests failed
    --- FAIL: TestParallel/subquery_retry_multinode (33.11s)
    	testing.go:730: race detected during execution of test
	testing.go:730: race detected during execution of test

```

--- a/pkg/storage/replica_command.go
+++ b/pkg/storage/replica_command.go
@@ -1072,6 +1072,12 @@ func updateRangeDescriptor(
 func (s *Store) AdminRelocateRange(
 	ctx context.Context, rangeDesc roachpb.RangeDescriptor, targets []roachpb.ReplicationTarget,
 ) error {
+
+	// Deep-copy the Replicas slice (in our shallow copy of the RangeDescriptor)
+	// since we'll mutate it in the loop below.
+	rangeDesc.Replicas = append([]roachpb.ReplicaDescriptor(nil), rangeDesc.Replicas...)
+	startKey := rangeDesc.StartKey.AsRawKey()
+
 	// Step 1: Compute which replicas are to be added and which are to be removed.
 	//
 	// TODO(radu): we can't have multiple replicas on different stores on the
@@ -1128,7 +1134,7 @@ func (s *Store) AdminRelocateRange(
 
 	transferLease := func() {
 		if err := s.DB().AdminTransferLease(
-			ctx, rangeDesc.StartKey.AsRawKey(), targets[0].StoreID,
+			ctx, startKey, targets[0].StoreID,
 		); err != nil {
 			log.Warningf(ctx, "while transferring lease: %s", err)
 		}
@@ -1146,10 +1152,6 @@ func (s *Store) AdminRelocateRange(
 	storeList, _, _ := s.allocator.storePool.getStoreList(rangeDesc.RangeID, storeFilterNone)
 	storeMap := storeListToMap(storeList)
 
-	// Deep-copy the Replicas slice (in our shallow copy of the RangeDescriptor)
-	// since we'll mutate it in the loop below.
-	desc := rangeDesc
-	desc.Replicas = append([]roachpb.ReplicaDescriptor(nil), desc.Replicas...)
 	rangeInfo := RangeInfo{Desc: &rangeDesc}
 
 	// Step 2: Repeatedly add a replica then remove a replica until we reach the
@@ -1204,7 +1206,7 @@ func (s *Store) AdminRelocateRange(
 				StoreID: targetStore.StoreID,
 			}
 			if err := s.DB().AdminChangeReplicas(
-				ctx, rangeDesc.StartKey.AsRawKey(), roachpb.ADD_REPLICA, []roachpb.ReplicationTarget{target},
+				ctx, startKey, roachpb.ADD_REPLICA, []roachpb.ReplicationTarget{target},
 			); err != nil {
 				returnErr := errors.Wrapf(err, "while adding target %v", target)
 				if !canRetry(err) {
@@ -1242,7 +1244,7 @@ func (s *Store) AdminRelocateRange(
 			// the leaseholder now, so we can always transfer the lease there.
 			transferLease()
 			if err := s.DB().AdminChangeReplicas(
-				ctx, rangeDesc.StartKey.AsRawKey(), roachpb.REMOVE_REPLICA, []roachpb.ReplicationTarget{target},
+				ctx, startKey, roachpb.REMOVE_REPLICA, []roachpb.ReplicationTarget{target},
 			); err != nil {
 				log.Warningf(ctx, "while removing target %v: %s", target, err)
 				if !canRetry(err) {

